### DISCUSSION TITLE
Ability to create Data Access modules in different namespaces

### DISCUSSION THREAD

**Discussion by rachit1arora** [fybrik/fybrik#774] [Ideas]
A toggle option as a configuration parameter for fybrik which can determine where data access module can be deployed. Whether it can be deployed in a pre configured namespace or it can be deployed in a new namespace 

Motivation for the requirment: If fybrik is being used in a multi tenant kind of environment where multiple tenants will be creating fybrik applications and for each application we will be starting a data access module the resources used in each data access module should be isolated from each other and one way to isolate them is by creating them in separate tenant specific namespaces and each namespace can have  some network polices also added for achieving right isolation. 

In current form of fybrik it can be data access modules can only be created in a single namespace. 

Some additional thought for this requirement
- Blueprint CRs to be created in one system namespace and not in the namespace where data access module will actually be started
- Blueprint Controller can rely on another component to get the require role and role binding to deploy resources in a new namespace which is created for the data access module
- We can use either a separate spec attribute or a label in fybrik application which can be used to determine the name of new namespace in which data access module need to be started. 

@roee88 @laurieaw63 Please add your thoughts and comments on this requirement

**Comment by roee88** [comment]
> Blueprint Controller can rely on another component to get the require role and role binding to deploy resources in a new namespace which is created for the data access module

This component also creates the namespace right? So we need code in blueprint controller to create/delete namespace using such a component and wait for completion. My assumption is that we will create a `NamespaceManager` interface for creation/deletion. Is that sufficient? 

> We can use either a separate spec attribute or a label in fybrik application which can be used to determine the name of new namespace in which data access module need to be started.

FYI previously we used a hash over the FybrikApplication namespace and name so the user had no power to deploy to any explicit namespace. 

Just so we won't forget - a separate item to address at a later point is to validate that deployed charts only apply resources to the target namespace (not to other namespaces and not any cluster scoped resources).

Should we move this to an issue?

**Reply by laurieaw63** [reply]
> We can use either a separate spec attribute or a label in fybrik application which can be used to determine the name of new namespace in which data access module need to be started.

> FYI previously we used a hash over the FybrikApplication namespace and name so the user had no power to deploy to any explicit namespace.

When the FybrikApplication is coming from a trusted source, we need a way to identify the namespace in the FybrkApplication, the name cannot be derived off the FybrikApplication namespace. In some environments the designated namespaces for the data access module will be pre created by an administrator.

**Reply by shlomitk1** [reply]
If I am not mistaken there is a limit on how many namespaces one can create in a single cluster. Wouldn't this become an issue for running tens of thousands of FybrikApplications having each created a new namespace for orchestrated modules?

**Reply by rachit1arora** [reply]
There will be another component which will watch for inactivity of fybrik application. That component will help to limit the number of fybrik application (which will results in namespace) that will run.

**Reply by simanadler** [reply]
> There will be another component which will watch for inactivity of fybrik application. That component will help to limit the number of fybrik application (which will results in namespace) that will run.

Perhaps this component should run as part of the control plane?  Maybe it should be implemented as a module?  @roee88 your thoughts?

**Reply by roee88** [reply]
> When the FybrikApplication is coming from a trusted source, we need a way to identify the namespace in the FybrkApplication, the name cannot be derived off the FybrikApplication namespace. In some environments the designated namespaces for the data access module will be pre created by an administrator.

@laurieaw63 Then the "trusted model" must be well documented. I'm not sure if we want to keep it as the only model or not. For example, we already treat context fields in the FybrikApplication as trusted (so a system is expected to use something like OPA gatekeeper or kyverno to limit context values by namespace/user). @rachit1arora opinions? 

> If I am not mistaken there is a limit on how many namespaces one can create in a single cluster. Wouldn't this become an issue for running tens of thousands of FybrikApplications having each created a new namespace for orchestrated modules?

@shlomitk1  I'm not aware of such limit. Do you have a reference to share? I can't see it in https://kubernetes.io/docs/setup/best-practices/cluster-large/. Anyway long term the goal should be shared resources whenever possible with namespaces indented for tenancy isolation. Supporting "too many" tenants in a single cluster is not a fybrik issue. 

> Perhaps this component should run as part of the control plane? Maybe it should be implemented as a module? @roee88 your thoughts?

@simanadler  I don't think this is in scope for the control plane, nor how it can be.

**Reply by shlomitk1** [reply]
@roee88 I heard a lecture about limitations on a number of namespaces that originate from etcd storage limits but it may be farfetched...
